executable-spec/src/Cardano/Ledger/Spec/STS/Update/Hupdate.hs

```diff
@@ -78,11 +98,59 @@ instance  STS (HUPDATE hashAlgo) where
           -- Add newly revealed (but stable) SIPs to the active sips. Note that
           -- we place these new sips as arguments of the left hand side of the
           -- 'Map.union', since this operation is left biased.
-          asips' = (wrsips ▷<= (slot -. (2 *. k))) `Map.union` asips
+          asips' = ( Map.mapWithKey  -- update asips slot with voting period end slot
+                       (\sph _ -> slot `addSlot` (votPeriodEnd sph))
+                       (wrsips ▷<= (slot -. (2 *. k)))
+
+                   )
+                   `Map.union`
+                   asips
+
+          votPeriodEnd siphash =  Data.vpDurationToSlotCnt
+                                  $ Data.votPeriodDuration
+                                  . Data.metadata
+                                  . Data.sipPayload
+                                  $ (sipdb!siphash)
           -- exclude old revealed SIPs
           wrsips' = dom asips' ⋪ wrsips
 
-      pure $ St { wrsips = wrsips'
-                , asips = asips'
-                }
+          -- Calculate SIPHashes to be tallied
+          toTally = (map fst)
+                      $ Map.toList
+                      $ (asips' ▷<= (slot -. (2 *. k)))
+
+          -- Prune asips, in order to avoid re-tallying of the same SIP
+          --asips'' = (dom toTally) ⋪ asips'
+          asips'' = Map.filterWithKey (\sh _ -> sh `notElem` toTally) asips'
```

Comment:
Because I was getting a compilation error. Now I have found it. It needs to be
```
asips'' = (Set.fromList toTally) ⋪ asips'
```
because `toTally` is a list

Reply:
Done OK